src/layout/CustomButton/CustomButtonComponent.tsx

```diff
@@ -0,0 +1,202 @@
+    previousPage: promisify(navigateToPreviousPage),
+    navigateToPage: promisify<ClientActionHandlers['navigateToPage']>(async ({ page }) => navigateToPage(page)),
+  };
+
+  const handleClientAction = async (action: CBTypes.ClientAction) => {
+    if (isSpecificClientAction('navigateToPage', action)) {
+      return await frontendActions[action.name](action.metadata);
+    }
+    await frontendActions[action.name]();
+  };
+
+  return {
+    handleClientActions: async (actions) => {
+      for (const action of actions) {
+        await handleClientAction(action);
+      }
+    },
+    handleDataModelUpdate: async (updatedDataModels) => {
+      const currentDataModelUpdates = currentDataModelGuid && updatedDataModels[currentDataModelGuid];
+      if (currentDataModelUpdates) {
+        dispatch(FormDataActions.fetchFulfilled({ formData: flattenObject(currentDataModelUpdates) }));
+      }
+    },
+  };
+}
+
+type PerformActionMutationProps = {
+  action: CBTypes.CustomAction;
+  buttonId: string;
+};
+
+type UsePerformActionMutation = {
+  mutation: UseMutationResult<ActionResult>;
+  performAction: (props: PerformActionMutationProps) => Promise<void>;
+};
+
+type PerformActionMutationError = {
+  response: {
+    data: {
+      error: {
+        message: string;
+        code: string;
+        metadata: Record<string, unknown>;
+      };
+    };
+  };
+};
+
+function usePerformActionMutation(): UsePerformActionMutation {
+  const { doPerformAction } = useAppMutations();
+  const { partyId, instanceGuid } = useNavigationParams();
+
+  const mutation = useMutation({
+    mutationFn: async ({ action, buttonId }: PerformActionMutationProps) => {
+      if (!instanceGuid || !partyId) {
+        throw Error('Cannot perform action without partyId and instanceGuid');
+      }
+      return doPerformAction.call(partyId, instanceGuid, { action: action.name, buttonId });
+    },
+  });
+  const { handleClientActions, handleDataModelUpdate } = useHandleClientActions();
+
+  return {
+    mutation,
+    /**
+     * We wrap the mutation in a promise to make it possible to await both the mutation,
+     * and the side effects that are run in the onSuccess or onError callbacks.
+     */
+    performAction: ({ action, buttonId }: PerformActionMutationProps) =>
+      new Promise((resolve) =>
+        mutation.mutate(
+          { action, buttonId },
+          {
+            onSuccess: async (data) => {
+              if (data.updatedDataModels) {
+                await handleDataModelUpdate(data.updatedDataModels);
+              }
+              if (data.clientActions) {
+                await handleClientActions(data.clientActions);
+              }
+
+              resolve();
+            },
+            onError: async (error: PerformActionMutationError) => {
+              if (error?.response?.data?.error?.message !== undefined) {
+                toast(error?.response?.data?.error?.message, { type: 'error' });
```

Comment:
I would assume this should be using a `<Lang id={error.response.data.error.message} />` as well, so that returned errors can be text resource keys (and thus translated).